cmd/internal/cli/x509.go

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2022, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package cli
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ocsp"
+)
+
+var errFailedToDecodePEM = errors.New("failed to decode PEM")
+
+// loadCertificate returns the certificate read from path.
+func loadCertificate(path string) (*x509.Certificate, error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	p, _ := pem.Decode(b)
+	if p == nil {
+		return nil, errFailedToDecodePEM
+	}
+
+	return x509.ParseCertificate(p.Bytes)
+}
+
+// loadCertificatePool returns the pool of certificates read from path.
+func loadCertificatePool(path string) (certPool, error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	pool := make(map[string]*x509.Certificate)
+
+	for rest := bytes.TrimSpace(b); len(rest) > 0; {
+		var p *pem.Block
+
+		if p, rest = pem.Decode(rest); p == nil {
+			return nil, errFailedToDecodePEM
+		}
+
+		cert, err := x509.ParseCertificate(p.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		pool[string(cert.SubjectKeyId)] = cert
+	}
+
+	return pool, nil
+}
+
+// certPool is needed because *x509.x509CertPool does not support listing of Certificates.
+// Without listing, it is impossible to perform Revocation Checking.
+type certPool map[string]*x509.Certificate
```

Comment:
I'm not sure I understand why we can't do revocation listing without defining this type?

The logic in `internal/app/singularity` is calling [Certificate.Verify](https://pkg.go.dev/crypto/x509#Certificate.Verify) ([here](https://github.com/fnikolai/singularity-x509/blob/f3d5b75b0ac3023c445b246f4ceb4f3a7909cb33/internal/app/singularity/verify.go#L168)), which will return one or more certificate chains based on the certificate and pool(s) it is passed.

When `--ocsp-verify` is specified, I think what we'll need to do is add on to that by walk the certificate chain(s) and do OCSP verification (call `queryOCSP`) on each certificate in the chain. Does that make sense?